[Docs] Rewrite run methods and deployment docs (#945)
mckinsey/vizro: vizro-core/docs/pages/user-guides/run.md

# How to run your dashboard in development

Typically when you create a dashboard, there are two distinct stages:

> 1. Development. This is when you build your app. You make frequent changes to your code and want to see how the dashboard looks after each change. You may or may not want to make the dashboard accessible to other people.

Author: ```suggestion
1. Development. This is when you build your app. You make frequent changes to your code and want to see how the dashboard looks after each change. At this point, you may, or may not, want to make the dashboard visible to other people.
```

Author: I've written round the word "accessibility" to avoid ambiguity around the term.

Author: Please can you take the changes over to the deployment page too.

> 1. Production. When you complete development of your app, you _deploy_ it to production. The dashboard should be accessible to other people.

Author: ```suggestion
1. Production. When you complete development of your app, you _deploy_ it to production. At this point, other people can access your dashboard.
```

> This page describes methods to run your dashboard _in development_. When you are ready to deploy your app to production then you should read our [guide to deployment](deploy.md).

Author: ```suggestion
This page describes methods to run your dashboard _in development_. When you are ready to deploy your app to production, read our [guide to deployment](deploy.md).
```

Vizro is built on top of [Dash](https://dash.plotly.com/), which itself uses [Flask](https://flask.palletsprojects.com/). Most of our guidance on how to run a Vizro app in development or production is very similar to guidance on Dash and Flask.

> ??? note

Author: ```suggestion
!!! note
```

Author: I don't think it's necessary to hide this behind a click.

    There are many possible workflows depending on your requirements. The above describes a simple workflow that applies to many people but is not suitable for everyone. For example:

    - If you are the only user of your app then the process is often simpler since you might never want to deploy to production.
    - If there are multiple people involved with development then you will need some way to coordinate code changes, such as a [GitHub repository](https://github.com/) or [Hugging Face Space](https://huggingface.co/spaces).

>     - You might want to make your app accessible to only some people using _authentication_.

Author: ```suggestion
    - You might want to use _authentication_ to restrict access to your app.
```

## Jupyter

If you develop in a Jupyter notebook or JupyterLab then you should use exactly the [same code as above](#development-server):

```python
from vizro import Vizro
import vizro.plotly.express as px
import vizro.models as vm

iris = px.data.iris()

page = vm.Page(
    title="My first page",
    components=[
        vm.Graph(figure=px.scatter(iris, x="sepal_length", y="petal_width", color="species")),
    ],
)

dashboard = vm.Dashboard(pages=[page])

Vizro().build(dashboard).run()
```

This runs Vizro using the Flask development server and shows the resulting dashboard inline in your notebook. You can change where the dashboard appears with the [`jupyter_mode` argument](https://dash.plotly.com/dash-in-jupyter). For example, `run(jupyter_mode="external")` provides a link to open the dashboard in a new window.

!!! note "Reloading and debugging"
    Code reloading and hot reloading do not work in Jupyter. There are two alternatve methods to reload the dashboard after you change your code:

    - Restart the Jupyter kernel and re-run your notebook.

>     - Add a cell containing `from vizro import Vizro; Vizro._reset()` to the top of your notebook and re-run it. With this method, there is no need to restart the Jupyter kernel.

Author: ```suggestion
    - Add a cell containing `from vizro import Vizro; Vizro._reset()` to the top of your notebook and re-run it before you re-run your code. With this method, there is no need to restart the Jupyter kernel.
```

## PyCafe

> The easiest way to run a dashboard is to work on the code live on [PyCafe](https://py.cafe/).

Author: Maybe put this first then?

Reviewer: I left it there as the local version is the most common, so they would both deserve first place in some sense :)
